File: src/main/scala/io/koju/autopos/config/AsyncConfiguration.scala
package io.koju.autopos.config

import java.util.concurrent.Executor

import io.koju.autopos.async.ExceptionHandlingAsyncTaskExecutor
import org.slf4j.{Logger, LoggerFactory}
import org.springframework.aop.interceptor.{AsyncUncaughtExceptionHandler, SimpleAsyncUncaughtExceptionHandler}
import org.springframework.context.annotation.{Bean, Configuration}
import org.springframework.scheduling.annotation.{AsyncConfigurer, EnableAsync, EnableScheduling}
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor

@Configuration
@EnableAsync
@EnableScheduling
class AsyncConfiguration(private val jHipsterProperties: AutoposProperties)
  extends AsyncConfigurer {

  final private val log: Logger = LoggerFactory.getLogger(classOf[AsyncConfiguration])

  @Bean(name = Array("taskExecutor"))
  override def getAsyncExecutor: Executor = {
    log.debug("Creating Async Task Executor")
    val executor: ThreadPoolTaskExecutor = new ThreadPoolTaskExecutor
    executor.setCorePoolSize(jHipsterProperties.getAsync.getCorePoolSize)
    executor.setMaxPoolSize(jHipsterProperties.getAsync.getMaxPoolSize)
    executor.setQueueCapacity(jHipsterProperties.getAsync.getQueueCapacity)
    executor.setThreadNamePrefix("autopos-Executor-")
    new ExceptionHandlingAsyncTaskExecutor(executor)
  }

  override def getAsyncUncaughtExceptionHandler: AsyncUncaughtExceptionHandler =
    new SimpleAsyncUncaughtExceptionHandler
}

File: src/main/kotlin/com/kapilkoju/autopos/config/AsyncConfiguration.kt
package com.kapilkoju.autopos.config

import io.github.jhipster.async.ExceptionHandlingAsyncTaskExecutor
import io.github.jhipster.config.JHipsterProperties
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.aop.interceptor.SimpleAsyncUncaughtExceptionHandler
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.scheduling.annotation.AsyncConfigurer
import org.springframework.scheduling.annotation.EnableAsync
import org.springframework.scheduling.annotation.EnableScheduling
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor
import java.util.concurrent.Executor

@Configuration
@EnableAsync
@EnableScheduling
class AsyncConfiguration(val jHipsterProperties: JHipsterProperties) : AsyncConfigurer {

    final private val log: Logger = LoggerFactory.getLogger(AsyncConfiguration::class.java)

    @Bean(name = ["taskExecutor"])
    override fun getAsyncExecutor(): Executor {
        log.debug("Creating Async Task Executor")
        val executor = ThreadPoolTaskExecutor()
        executor.corePoolSize = jHipsterProperties.async.corePoolSize
        executor.maxPoolSize = jHipsterProperties.async.maxPoolSize
        executor.setQueueCapacity(jHipsterProperties.async.queueCapacity)
        executor.threadNamePrefix = "autopos-Executor-"
        return ExceptionHandlingAsyncTaskExecutor(executor)
    }

    override fun getAsyncUncaughtExceptionHandler() = SimpleAsyncUncaughtExceptionHandler()
}